### PR TITLE
Kotlin inline functions support: attach mode

### PR DESCRIPTION
**Pull request by zuevmaxim**

### CODE REVIEW

File: src/debugInfoAccess.h [hunk 1]
```diff
@@ -31,5 +36,25 @@ class DebugInfoAccessJVMTI : public DebugInfoAccess {
     jvmtiError GetLocalVariableTable(jvmtiEnv* env, jmethodID method, jint* entry_count_ptr, jvmtiLocalVariableEntry** table_ptr, bool* deallocate);
 };
 
+class DebugInfoAccessFromClassFiles : public DebugInfoAccess {
+private:
```
Comment (nikita-nazarov) [inline src/debugInfoAccess.h:40]: The tabulation of `private` and `public` here is not on par with the tabulation in the class declaration above.

File: src/debugInfoAccess.cpp [hunk 2]
```diff
@@ -77,3 +83,253 @@ jvmtiError DebugInfoAccessJVMTI::GetSMAPEntry(jvmtiEnv* jvmti, jclass klass, SMA
...
+    std::string name;
+    err = getClassName(jvmti, klass, name);
+    if (err != JVMTI_ERROR_NONE) return err;
+    _mutex.lock();
+    std::map<std::string, int>::iterator index_it = _seen_classes.find(name);
+    if (index_it == _seen_classes.end()) {
+        _mutex.unlock();
+        return JVMTI_ERROR_ABSENT_INFORMATION;
+    }
+    std::map<int, std::map<std::string, std::vector<jvmtiLocalVariableEntry> > >::iterator it = DebugInfoAccessFromClassFiles::_local_variables.find(index_it->second);
```
Comment (nikita-nazarov) [inline src/debugInfoAccess.cpp:119]: This line is way too long

File: src/debugInfoAccess.cpp [hunk 3]
```diff
@@ -77,3 +83,253 @@ jvmtiError DebugInfoAccessJVMTI::GetSMAPEntry(jvmtiEnv* jvmti, jclass klass, SMA
...
+
+      for (int i = 0; i < attributes_count; i++) {
+          u16 attribute_name_index = get16();
+
+          Constant* attribute_name = cpool[attribute_name_index];
+          u32 attribute_length = get32();
+          if (scope == SCOPE_METHOD && attribute_name->equals("Code", 4)) {
+              analyseCode(cpool, method_id);
+          } else if (scope == SCOPE_REWRITE_CODE && attribute_name->equals("LocalVariableTable", 18)) {
+              std::vector <jvmtiLocalVariableEntry> localVariables = analyseLocalVariablesTable(cpool);
```
Comment (nikita-nazarov) [inline src/debugInfoAccess.cpp:220]: Extra whitespace `std::vector <jvmtiLocalVariableEntry>`

File: src/debugInfoAccess.cpp [hunk 4]
```diff
@@ -77,3 +83,253 @@ jvmtiError DebugInfoAccessJVMTI::GetSMAPEntry(jvmtiEnv* jvmti, jclass klass, SMA
...
+        jboolean is_modifiable;
+        bool is_kotlin_file;
+        if (jvmti->IsModifiableClass(classes[i], &is_modifiable) == 0 && is_modifiable == JNI_TRUE) {
+            if (IsKotlinClass(jvmti,  classes[i], &is_kotlin_file) == 0 && !is_kotlin_file) continue;
+            char* signature;
+            if (jvmti->GetClassSignature(classes[i], &signature, NULL) == 0) {
+                if (signature[0] == 'L') {
+                    if (!shouldAnalyseClass(signature + 1)) continue;
+                    std::string class_name(signature + 1, strlen(signature) - 2);
+                    if (_seen_classes.find(class_name) != _seen_classes.end()) continue;
```
Comment (nikita-nazarov) [inline src/debugInfoAccess.cpp:315]: The `find` call is not under a lock here.

File: src/debugInfoAccess.h [hunk 5]
```diff
@@ -31,5 +36,25 @@ class DebugInfoAccessJVMTI : public DebugInfoAccess {
...
+  static Mutex _mutex;
+  static std::map<int, std::map<std::string, std::vector<jvmtiLocalVariableEntry> > > _local_variables;
```
Comment (nikita-nazarov) [inline src/debugInfoAccess.h:42]: It looks like it could've been a vector, but map is fine too

Comment (nikita-nazarov) [reply src/debugInfoAccess.h:42]: By the way why map and not unordered_map? Especially for the second type parameter it should be better.

**Review by nikita-nazarov — APPROVED** [review]
Hi Maxim! Everything is cool except for a couple of minor comments.

File: src/debugInfoAccess.h (hunk 5 above)
Comment (zuevmaxim) [reply src/debugInfoAccess.h:42]: You are right, I'll replace the map with a vector

Comment (zuevmaxim) [reply src/debugInfoAccess.h:42]: We are unable to use unordered_map because the version of C++ STL that is used in this project does not include it

**Pull request closed without merging**